Add Bulk controller and unit tests

Shiranuit opened the pull request:
<!--
  This template is optional.
  It simply serves to provide a guide to allow a better review of pull requests.
-->

<!--
  IMPORTANT
  Don't forget to add the corresponding "changelog:xxx" label to your PR.
  This is part of our release process in order to generate the change log.
-->


## What does this PR do ?
<!-- Please fill this section -->
Add a BulkController with methods:
- ImportAsync
- mWriteAsync
- WriteAsync

Add unit tests for the BulkController
<!--
  Please include a summary of the change and which issue is fixed.
  Please also include relevant motivation and context.
  List any dependencies that are required for this change.
-->

### How should this be manually tested?

<!--
  Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.
  Please also list any relevant details for your test configuration
-->
Run the following command : `dotnet test`

### Boyscouts

Removed the method HandleRefreshOption from DocumentController to use
`api.QueryAsync(JObject [, QueryOptions])`

Replace `if (refresh) expected.Add("refresh", "wait_for");` to use `api.Verify(JObject [, QueryOptions])` instead to make modification to the refresh API option easier by modifying the QueryOptions.AddToQuery method behaviour.

Commit:
add Bulk Controller methods [import, mWrite, write] and Unit test for [import, mWrite]

Commit:
removing wrong BulkController file in Kuzzle.Tests

Commit:
Fix BulkController unit tests for methods [mWrite, write]

Commit:
Merge branch '0-dev' into bulk-controller

Commit:
Add summary for each BulkController method

Commit:
remove unused import

Commit:
rename unit tests methods being wrongly named

scottinet commented on Kuzzle.Tests/API/Controllers/BulkControllerTest.cs:
```diff
@@ -0,0 +1,115 @@
+    [Theory]
+    [InlineData(false)]
+    [InlineData(true)]
+    public async void mWriteAsyncTestSuccess(bool notify) {
```
Public methods should be capitalized.

```suggestion
    public async void MWriteAsyncTestSuccess(bool notify) {
```

Shiranuit replied:
At the time, I didn't know what to do with this special case.

I've changed it for a Capital letter.

scottinet commented on Kuzzle/API/Controllers/BulkController.cs:
```diff
@@ -0,0 +1,86 @@
+    /// <summary>
+    /// Create or replace multiple documents directly into the storage engine.
+    /// </summary>
+    public async Task<JObject> mWriteAsync(
```
Public methods should be capitalized.

```suggestion
    public async Task<JObject> MWriteAsync(
```

scottinet commented on Kuzzle/API/Controllers/BulkController.cs:
```diff
@@ -0,0 +1,86 @@
+        string collection,
+        string documentContent,
+        string documentId = null,
+        string refresh = null,
```
All SDKs replaces this API option with a boolean `WaitForRefresh` one. Under the hood, if set to true, the request is sent with a `refresh: 'wait_for'` value set in the JSON payload.
(check the document controller for instance)

In the next major version of Kuzzle, the API option `refresh: 'wait_for'` will probably be deprecated in favor of a `waitForRefresh: <boolean>` option.


(I won't report this issue more than once but it must be fixed everywhere obviously)

Shiranuit replied:
This should be resolved now, I've changed everything related to the refresh option.
now there is a QueryUtils class with static methods [HandleRefreshOption, HandleNotifyOption] that should do the job.

Also, I've changed every Controller's methods that have a refresh parameter to use QueryUtils.HandleRefreshOption instead and avoid copy paste.

Same thing for every unit tests, I replaced every `if (refresh) expected.Add("refresh", "wait_for");` with `QueryUtils.HandleRefreshOption(expected, refresh)` instead, this would be easier if the actual API option `refresh: "wait_for"` needs to be deprecated.

scottinet replied:
I agree with all these changes but the last one: unit tests should add the property themselves without resorting to HandleRefreshOption. That way, we can make sure that JSON requests are properly built, testing the new Handlexxx methods along the way.

Shiranuit replied:
This has been changed, now there is a QueryUtils inside Kuzzle.Tests

scottinet commented on Kuzzle/API/Controllers/BulkController.cs:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    /// <summary>
+    /// Create or replace a document directly into the storage engine.
```
```suggestion
    /// Creates or replaces a document directly into the storage engine.
```

(the documentation is wrong on that regard, I'll write a PR to fix it)

Shiranuit replied:
This typo should be fixed

scottinet commented on Kuzzle/API/Controllers/BulkController.cs:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    /// <summary>
+    /// Create or replace multiple documents directly into the storage engine.
```
```suggestion
    /// Creates or replaces multiple documents directly into the storage engine.
```

Shiranuit replied:
This typo should be fixed

scottinet commented on Kuzzle.Tests/Kuzzle.Tests.csproj:
```diff
@@ -28,4 +28,7 @@
     <Folder Include="API\Controllers\" />
   </ItemGroup>
+  <ItemGroup>
+    <Compile Remove="API\Controllers\BulkController.cs" />
```
should be removed: the previous item group already includes the entire API/Controllers folder

Commit:
Fix public methods not being capitalized

Commit:
- Move HandleRefreshOption method from DocumentController to QueryUtils.cs,
convert it to static method to be reuse and avoid copy paste

- Change every 'if (refresh) expected.add("refresh", "wait_for")' to use
QueryUtils.HandleRefreshOption (easier to change if the API option 'refresh: "wait_for"' is going to be deprecated)

- fix typo in MWriteAsync summary

- fix typo in WriteAsync summary

Commit:
remove wrong ItemGroup in Kuzzle.Tests/Kuzzle.Tests.csproj

Commit:
add the BulkController to the Kuzzle Class

scottinet commented on Kuzzle.Tests/API/Controllers/BulkControllerTest.cs:
```diff
@@ -52,32 +55,34 @@ public class BulkControllerTest {
+    [InlineData("", "documentId", false, false)]
+    [InlineData("Some input", "foobar", true, false)]
+    [InlineData("Some input", "foobar", false, true)]
+    public async void WriteAsyncTestSuccess(string documentInput, string documentId, bool notify, bool refresh = false) {
```
(nitpicking) you don't need to default the `refresh` argument

Shiranuit replied:
True, the refresh argument has no default value now.

Commit:
remove passing default argument to tests with default parameter

Commit:
change QueryUtils.HandleRefreshOption in Kuzzle.Tests to use it's own QueryUtils.HandleRefreshOption to test the one in Kuzzle

benoitvidis reviewed (changes requested):
I am sorry I missed the first discussion, but I am really not found of utility classes. 
They usually tend to end-up being a big mess of totally unrelated static functions.

I would rather move the two little helpers code (actually possibly only one of them) down to the `api.QueryAsync` instead of sharing it between the controllers.

benoitvidis commented on Kuzzle/Utils/QueryUtils.cs:
```diff
@@ -0,0 +1,19 @@
+    public static void HandleNotifyOption(JObject query, bool notify) {
+      if (notify) {
+        query.Add("notify", true);
+      }
```
If we happen to keep this util class, couldn't this method call just be replaced by `query.Add("notify", notify)`? 
I don't see any situation where setting it to false in the `JObject` could cause any harm.

Shiranuit replied:
This is true, so I've changed it to do `query.Add("notify", notify)` instead.

Commit:
Remove refresh being an optionnal argument

Shiranuit commented:
> I am sorry I missed the first discussion, but I am really not found of utility classes.
> They usually tend to end-up being a big mess of totally unrelated static functions.
> 
> I would rather move the two little helpers code (actually possibly only one of them) down to the `api.QueryAsync` instead of sharing it between the controllers.

In fact this could not be done if this has to be tested, the QueryAsync method may not contain an optional argument because she is Mocked and the MOQ framework uses an Expression Tree to redefine the behaviour of this method 
(for example in the api.SetResult method to redefine the return value of api.QueryAsync).

In this case if the QueryAsync method had an optional argument this would have triggered the error `CS0854: An expression tree may not contain a call or invocation that uses optional arguments`.

Commit:
HandleNotifyOption remove If statement before adding the notify option to query

scottinet commented on Kuzzle/Utils/QueryUtils.cs:
```diff
@@ -0,0 +1,17 @@
+      }
+    }
+
+    public static void HandleNotifyOption(JObject query, bool notify) {
```
This method adds a useless level of abstraction over a one-liner.

scottinet commented on Kuzzle.Tests/Utils/QueryUtils.cs:
```diff
@@ -0,0 +1,17 @@
+﻿using Newtonsoft.Json.Linq;
+
+namespace Kuzzle.Tests.Utils {
```
Too easy to have a name collision, with tests unexpectedly using the KuzzleSdk.Utils domain instead of this one.

scottinet commented on Kuzzle.Tests/Utils/QueryUtils.cs:
```diff
@@ -0,0 +1,17 @@
+      }
+    }
+
+    public static void HandleNotifyOption(JObject query, bool notify) {
```
This method is unneeded.

scottinet commented on Kuzzle/Utils/QueryUtils.cs:
```diff
@@ -0,0 +1,17 @@
+﻿using Newtonsoft.Json.Linq;
+
+namespace KuzzleSdk.Utils {
```
After reading @benoitvidis 's review I actually agree: this needs to be isolated in its own namespace, such as `KuzzleSdk.Api.QueryOptions` for instance

Shiranuit replied:
I've changed how query options are handled.
Now QueryAsync takes an QueryOptions as optional parameter.

scottinet committed:
Merge branch '0-dev' into bulk-controller

Commit:
remove HandleNotifyOption method from QueryUtils

Commit:
Merge branch 'bulk-controller' of https://github.com/kuzzleio/sdk-csharp into bulk-controller

Shiranuit commented:
> > I am sorry I missed the first discussion, but I am really not found of utility classes.
> > They usually tend to end-up being a big mess of totally unrelated static functions.
> > I would rather move the two little helpers code (actually possibly only one of them) down to the `api.QueryAsync` instead of sharing it between the controllers.
> 
~~In fact this could not be done if this has to be tested, the QueryAsync method may not contain an optional argument because she is Mocked and the MOQ framework uses an Expression Tree to redefine the behaviour of this method
 (for example in the api.SetResult method to redefine the return value of api.QueryAsync).~~
 
~~In this case if the QueryAsync method had an optional argument this would have triggered the error CS0854: An expression tree may not contain a call or invocation that uses optional arguments.~~

Nevermind, I've found a way to do it.

scottinet commented on Kuzzle/Kuzzle.cs:
```diff
@@ -42,7 +43,7 @@ public interface IKuzzleApi {
     /// <returns>The query response.</returns>
     /// <param name="query">Kuzzle API query</param>
-    Task<Response> QueryAsync(JObject query);
+    Task<Response> QueryAsync(JObject query, QueryOptions options = null);
```
It's worse than a global utility class: this new signatures tells our users that a QueryOptions object can be passed to all queries but the handled option values are only useful for several API routes only. 

First: the `notify` option doesn't need a utility option, just pass it in the JSON body of the actions handling that option.
Second: the only option that requires a dedicated DRYifying function because the SDK has a different way to handle it than the API, is the `waitForRefresh` one. So all you need is your previous "HandleRefreshOption" function in a dedicated namespace and that is all. It's still a utility function, but instead of being made available in a static object global to the project, it's neatly stored in a dedicated and meaningful namespace.

Shiranuit replied:
@scottinet @Aschen @benoitvidis It seems that your not all agree on what has to be done for this case, it's easy to change really quick but everyone seems to have a different opinion on that.

So I will let that has it is for now.

benoitvidis replied:
I am sorry, I must have missed something, but why do we need a utility function at all?
Can't we just set the `waitForRefresh` like we do for `notify` from the controllers, and in `queryAsync`, tweak the json with something like this? :

```csharp
[..]
      if (query["waitForRefresh"] != null) {
        if (query["waitForRefresh"].ToObject<bool>()) {
          query["refresh"] = "wait_for";
        }
        query["waitForRefresh"].Remove();
      }
[..]
```

scottinet replied:
I guess this is a possibility, but my first reaction was to disagree with it, because we're letting controllers set what we know is an unsupported option, to let the QueryAsync correct it afterward.

That said... I didn't think of that approach, and might have its usefulness depending on how we plan to handle the refresh option in the future.
I'm confident that we'll change it to a waitForRefresh boolean option, as this makes a lot more sense for users, and is more on par with the remaining of our API.

So... if we do that, and if we drop the support for the Kuzzle API v1 in this SDK, then we'd simply need to remove this snippet, instead of changing all methods to get rid of that utility function.

At that point, I guess it's up to @Shiranuit.

Shiranuit replied:
So, I've changed everything another time to use the solution of @benoitvidis that I think is better and avoid massive modification to remove the utility method.

Commit:
Revert comit and move QueryUtils in Kuzzle.API.QueryOptions namespace

benoitvidis reviewed (commented):
I still don't fancy utility tools but I can live with it too.. switching to neutral mode.

benoitvidis reviewed (approved):
I still don't fancy utility tools but I can live with it too..

Commit:
Remove utility method for handling refresh option, QueryAsync now convert waitForRefresh: true to refresh: 'wait_for'

scottinet reviewed (changes requested):
you need to add a new unit test on the QueryAsync method checking that it handles the `waitForRefresh` replacement correctly

Commit:
Add waitForRefresh unit test in QueryAsync

Aschen committed:
Merge branch '0-dev' into bulk-controller

Aschen reviewed (changes requested):
Missing documentation

Commit:
Merge branch '0-dev' into bulk-controller

Commit:
Add documentation with snippet tests

Commit:
Merge branch 'bulk-controller' of https://github.com/kuzzleio/sdk-csharp into bulk-controller

Aschen commented on doc/1/controllers/bulk/import-async/index.md:
```diff
@@ -0,0 +1,52 @@
+
+## Return
+
+An object containing information about the import status for each document.
```
```suggestion
A `JObject` containing information about the import status for each document.
```

Aschen commented on doc/1/controllers/bulk/import-async/snippets/import-async.cs:
```diff
@@ -0,0 +1,15 @@
+try {
+    //await kuzzle.Auth.LoginAsync("local", JObject.Parse("{username:'foo',password:'bar'}"));
```
Useless comment

Aschen commented on doc/1/controllers/bulk/mwrite-async/index.md:
```diff
@@ -0,0 +1,57 @@
+
+This is a low level route intended to bypass Kuzzle actions on document creation, notably:
+
+check document validity,
```
```suggestion
 - check document validity,
```

Aschen commented on doc/1/controllers/bulk/mwrite-async/index.md:
```diff
@@ -0,0 +1,57 @@
+This is a low level route intended to bypass Kuzzle actions on document creation, notably:
+
+check document validity,
+add kuzzle metadata,
```
```suggestion
 - add kuzzle metadata,
```

Aschen commented on doc/1/controllers/bulk/mwrite-async/index.md:
```diff
@@ -0,0 +1,57 @@
+
+check document validity,
+add kuzzle metadata,
+trigger realtime notifications (unless asked otherwise)
```
```suggestion
 - trigger realtime notifications (unless asked otherwise)
```

Aschen commented on doc/1/controllers/bulk/import-async/index.md:
```diff
@@ -0,0 +1,52 @@
+    string index,
+    string collection,
+    JArray bulkData
+)
```
Should be in the upper line

Aschen commented on doc/1/controllers/bulk/mwrite-async/index.md:
```diff
@@ -0,0 +1,57 @@
+    JArray documents,
+    bool waitForRefresh = false,
+    bool notify = false
+)
```
Should be in the upper line

Aschen commented on doc/1/controllers/bulk/mwrite-async/index.md:
```diff
@@ -0,0 +1,57 @@
+
+## Return
+
+A `hits` array, containing the list of created documents, in the same order than the one provided in the query.
```
```suggestion
A `JObject` containing a `hits` array which represent the list of created documents, in the same order than the one provided in the query.
```

Aschen commented on doc/1/controllers/bulk/mwrite-async/snippets/mwrite-async.cs:
```diff
@@ -0,0 +1,6 @@
+try {
+    await kuzzle.Bulk.MWriteAsync("nyc-open-data", "yellow-taxi", JArray.Parse("[{_id: 'foo', body: {}}]"));
```
Please fold this line

Aschen commented on doc/1/controllers/bulk/write-async/index.md:
```diff
@@ -0,0 +1,55 @@
+    string documentId = null,
+    bool waitForRefresh = false,
+    bool notify = false
+)
```
Should be in the upper line

Aschen commented on doc/1/controllers/bulk/write-async/index.md:
```diff
@@ -0,0 +1,55 @@
+async Task<JObject> WriteAsync(
+    string index,
+    string collection,
+    JObject documentContent,
```
```suggestion
    JObject content,
```

Aschen commented on doc/1/controllers/bulk/write-async/index.md:
```diff
@@ -0,0 +1,55 @@
+| ----------------- | ------------------ | --------------------------- |
+| `index`           | <pre>string</pre>  | Index name                  |
+| `collection`      | <pre>string</pre>  | Collection name             |
+| `documentContent` | <pre>JObject</pre> | Document content to create. |
```
```suggestion
| `content` | <pre>JObject</pre> | Document content to create. |
```

Aschen commented on doc/1/controllers/bulk/write-async/index.md:
```diff
@@ -0,0 +1,55 @@
+
+## Return
+
+A `hits` array, containing the list of created documents, in the same order than the one provided in the query.
```
Wrong return type: https://docs.kuzzle.io/core/1/api/controllers/bulk/write/

Aschen commented on doc/1/controllers/bulk/write-async/snippets/write-async.cs:
```diff
@@ -0,0 +1,6 @@
+try {
+    await kuzzle.Bulk.WriteAsync("nyc-open-data", "yellow-taxi", JObject.Parse("{foo: 'bar'}"));
```
An example with a `_kuzzle_info` property in the body is more appropriate since this is one of this route purpose

Aschen reviewed (changes requested):
Good job :+1: 
Some remarks about the documentation

Commit:
Apply @Aschen requests

Commit:
Apply @Aschen requests

Commit:
Add JSON representation of JObject after Console.WriteLine

Commit:
Remove signature, remove sdk and version from *.yml, fix typo and fix regex write-async

Commit:
Fix typo

Aschen committed:
Update doc/1/controllers/bulk/mwrite-async/index.md
